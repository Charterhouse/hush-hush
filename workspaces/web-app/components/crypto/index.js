export { Encryptor } from './Encryptor'
export { Inviter } from './Inviter'
