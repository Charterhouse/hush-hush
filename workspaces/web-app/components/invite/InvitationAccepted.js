import React from 'react'
import Link from 'next/link'
import { Button } from 'semantic-ui-react'
import { FadingValueBox } from '../animations'
import { Green } from '../ui'

const InvitationAccepted = () => (
  <FadingValueBox>
    <div css={{ display: 'flex', width: '100%', flexFlow: 'column nowrap', alignItems: 'center' }}>
      <div css={{ width: '100%', textAlign: 'center' }}>
        You <Green>accepted</Green> the invitation!
      </div>
      <div css={{ width: '100%', textAlign: 'center', margin: '20px 0 20px 0' }}>
        Now the sender can start hushing with you. Maybe it is worth to remind him?
      </div>
      <Link href={{ pathname: '/' }}>
        <Button primary>Done</Button>
      </Link>
    </div>
  </FadingValueBox>
)

export { InvitationAccepted }
