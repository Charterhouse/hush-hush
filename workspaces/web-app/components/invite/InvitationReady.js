import React from 'react'
import Link from 'next/link'
import { Button } from 'semantic-ui-react'
import { FadingValueBox } from '../animations'
import { Connector } from '../identity'
import { Green } from '../ui'

const InvitationReady = ({ accept }) => (
  <FadingValueBox>
    <div css={{ display: 'flex', width: '100%', flexFlow: 'column nowrap', alignItems: 'center' }}>
      <div css={{ width: '100%', textAlign: 'center' }}>
        Invitation looks <Green>good!</Green>
      </div>
      <div css={{ width: '100%', textAlign: 'center', margin: '20px 0 20px 0' }}>
        Do you want to accept it?
      </div>
      <div css={{ display: 'flex', justifyContent: 'center', marginTop: '1rem', width: '100%' }}>
        <Connector onDone={accept}
          title='Accept...' />
        <Link href={{ pathname: '/' }}>
          <Button>Maybe later...</Button>
        </Link>
      </div>
    </div>
  </FadingValueBox>
)

export { InvitationReady }
