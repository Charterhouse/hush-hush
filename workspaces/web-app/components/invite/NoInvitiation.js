import React from 'react'
import { Red } from '../ui'

const NoInvitation = () => (
  <div css={{ display: 'flex', flexFlow: 'column nowrap', alignItems: 'center', width: '100%' }}>
    <div>
      <Red>Oops!!!</Red> I could not find any potential invitation in your url.
    </div>
    <div>
      Please check that you copied and pasted the whole url from your email.
    </div>
    <div css={{ marginTop: '20px' }}>
      Don't give up. Let's hush something!
    </div>
  </div>
)

export { NoInvitation }
