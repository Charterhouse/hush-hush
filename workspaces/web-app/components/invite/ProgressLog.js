import React from 'react'
import { FadingValueBox } from '../animations'

const ProgressLog = ({ status }) => (
  <FadingValueBox trigger={status}>
    <div css={{ width: '100%', textAlign: 'center', wordBreak: 'break-word' }}>
      {status}
    </div>
  </FadingValueBox>
)

export { ProgressLog }
