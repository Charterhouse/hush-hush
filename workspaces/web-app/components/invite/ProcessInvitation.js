import React from 'react'
import Link from 'next/link'
import { Inviter } from '../crypto'
import { Connector } from '../identity'
import { FadingValueBox } from '../animations'
import { Button } from 'semantic-ui-react'

import { Red, Green, Blue } from '../ui'

class ProcessInvitation extends React.Component {
  state = {
    status: 'checking your invitation...',
    inProgress: true,
    invitationOk: false,
    invitationAccepted: false
  }

  log = (...args) => {
    const status = args.map((a, i) => {
      if (a.startsWith('[green]')) {
        return <Green key={i}>{a.slice('[green]'.length)}</Green>
      } if (a.startsWith('[blue]')) {
        return <Blue key={i}>{a.slice('[blue]'.length)}</Blue>
      } if (a.startsWith('[red]')) {
        return <Red key={i}>{a.slice('[red]'.length)}</Red>
      } else {
        return a
      }
    })
    this.setState({ status })
  }

  onStatusChanged = (...args) => {
    this.log(...args)
  }

  checkInvitation = async () => {
    console.log(this.props.invitation)
    try {
      const { tag, symmetricKey, nonce } = Inviter.check(this.props.invitation)
      setTimeout(() => {
        this.setState({
          tag,
          symmetricKey,
          nonce,
          invitationOk: true,
          inProgress: false
        })
        // this.log('invitation looks ', '[green]good')
      }, 3000)
    } catch (e) {
      this.setState({ inProgress: true, status: e.message })
    }
  }

  accept = async telepathChannel => {
    try {
      this.setState({
        inProgress: true
      })
      this.log('accepting invitation...')
      const { tag, symmetricKey, nonce } = this.state
      await Inviter.accept({
        tag,
        symmetricKey,
        nonce,
        telepathChannel,
        onStatusChanged: this.onStatusChanged
      })
      setTimeout(() => {
        this.setState({
          inProgress: false,
          invitationAccepted: true
        })
      }, 2000)
    } catch (e) {
      this.setState({ inProgress: true, status: e.message })
    }
  }

  componentDidMount () {
    this.checkInvitation()
  }

  render () {
    if (this.state.inProgress) {
      return (
        <FadingValueBox trigger={this.state.status}>
          <div css={{ width: '100%', textAlign: 'center', wordBreak: 'break-word' }}>
            {this.state.status}
          </div>
        </FadingValueBox>
      )
    }
    if (this.state.invitationAccepted) {
      return (
        <FadingValueBox>
          <div css={{ display: 'flex', width: '100%', flexFlow: 'column nowrap', alignItems: 'center' }}>
            <div css={{ width: '100%', textAlign: 'center' }}>
              You <Green>accepted</Green> the invitation!
            </div>
            <div css={{ width: '100%', textAlign: 'center', margin: '20px 0 20px 0' }}>
              Now the sender can start hushing with you. Maybe it is worth to remind him?
            </div>
            <Link href={{ pathname: '/' }}>
              <Button primary>Done</Button>
            </Link>
          </div>
        </FadingValueBox>
      )
    }
    if (this.state.invitationOk) {
      return (
        <FadingValueBox>
          <div css={{ display: 'flex', width: '100%', flexFlow: 'column nowrap', alignItems: 'center' }}>
            <div css={{ width: '100%', textAlign: 'center' }}>
              Invitation looks <Green>good!</Green>
            </div>
            <div css={{ width: '100%', textAlign: 'center', margin: '20px 0 20px 0' }}>
              Do you want to accept it?
            </div>
            <div css={{ display: 'flex', justifyContent: 'center', marginTop: '1rem', width: '100%' }}>
              <Connector onDone={this.accept}
                title='Accept...' />
              <Link href={{ pathname: '/' }}>
                <Button>Maybe later...</Button>
              </Link>
            </div>
          </div>
        </FadingValueBox>
      )
    }
    return null
  }
}

export { ProcessInvitation }
