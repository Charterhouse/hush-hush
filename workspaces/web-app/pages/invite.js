import React from 'react'
import { PageCentered } from '@react-frontend-developer/react-layout-helpers'
import { FadingValueBox } from '../components/animations'

import { InviteHush } from '../components/invite'

const Invite = () => (
  <PageCentered css={{ color: 'white' }}>
    <div css={{ display: 'flex', flexFlow: 'column nowrap', alignItems: 'center', minHeight: '150px', maxWidth: '550px', width: '85%' }}>
      <FadingValueBox>
        <InviteHush />
      </FadingValueBox>
    </div>
  </PageCentered>
)

export default Invite
