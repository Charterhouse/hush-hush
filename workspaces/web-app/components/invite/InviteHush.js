import React, { useState, useEffect } from 'react'

import { NoInvitation } from './NoInvitiation'
import { ProcessInvitation } from './ProcessInvitation'

const InviteHush = () => {
  const [invitationOk, setInvitationOk] = useState(false)
  const [invitation, setInvitation] = useState(null)

  useEffect(() => {
    if (window.location.hash) {
      const invitation = window.location.hash.substring(1)
      console.log(invitation)
      setInvitationOk(true)
      setInvitation(invitation)
    }
  })

  if (!invitationOk) {
    return (
      <NoInvitation />
    )
  }

  return (
    <ProcessInvitation invitation={invitation} />
  )
}

export { InviteHush }
