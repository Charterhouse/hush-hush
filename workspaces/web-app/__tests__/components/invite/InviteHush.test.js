import React from 'react'
import { InviteHush } from '../../../components/invite'
import { render, waitForElement } from 'react-testing-library'

describe('InviteHush', () => {
  beforeEach(() => {
    console.log = jest.fn()
    jest.useFakeTimers()
  })

  afterEach(() => {
    console.log.mockRestore()
  })

  const renderWithHash = (hash = '') => {
    window.location.hash = hash
    return render(<InviteHush />)
  }

  it('renders the info when no invitation fragment is present in the url', () => {
    const { getByText } = renderWithHash()

    expect(getByText(/could not find any potential invitation in your url/i)).toBeInTheDocument()
  })

  it('display error message if invitation hash does not include key component', async () => {
    const { getByText } = renderWithHash('#tag')

    await waitForElement(() => getByText(/missing symmetric key component in the invitation/i))
  })

  it('display error message if invitation hash does not include nonce component', async () => {
    const { getByText } = renderWithHash('#tag.key')

    await waitForElement(() => getByText(/missing nonce component in the invitation/i))
  })

  it('informs the user that invitation looks good if the invitation hash includes all required components', async () => {
    const { getByText } = renderWithHash('#tag.key.nonce')
    jest.advanceTimersByTime(3000)

    await waitForElement(() => getByText(/invitation looks/i))
    expect(getByText(/good/i)).toBeInTheDocument()
  })

  it('informs the user that invitation looks good - snapshot', async () => {
    const { getByText, container } = renderWithHash('#tag.key.nonce')
    jest.advanceTimersByTime(3000)

    await waitForElement(() => getByText(/invitation looks/i))
    expect(container.firstChild).toMatchSnapshot()
  })
})
