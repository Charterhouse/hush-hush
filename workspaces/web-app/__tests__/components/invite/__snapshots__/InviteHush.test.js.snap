// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteHush informs the user that invitation looks good - snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  opacity: 1;
  -webkit-transition: opacity 1s ease-in-out 0s;
  transition: opacity 1s ease-in-out 0s;
}

<div
  class="emotion-0 emotion-1"
  opacity="1"
>
  <div
    css="[object Object]"
  >
    <div
      css="[object Object]"
    >
      Invitation looks 
      <span
        css="[object Object]"
      >
        good!
      </span>
    </div>
    <div
      css="[object Object]"
    >
      Do you want to accept it?
    </div>
    <div
      css="[object Object]"
    >
      <button
        class="ui button"
      >
        Maybe later...
      </button>
    </div>
  </div>
</div>
`;
